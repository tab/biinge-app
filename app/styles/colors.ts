const colors = {
  americanSilver: "#d1cecf",
  black: "#181818",
  darkBlue: "#2a3642",
  darkCandyAppleRed: "#9f0909",
  graniteGray: "#666666",
  gray: "#b5b5b5",
  grayDark: "#6e6969",
  gunmetal: "#2b2835",
  lotion: "#fafafa",
  orangeYellow: "#eec01e",
  palmLeaf: "#699d3a",
  raisinBlack: "#221f1f",
  white: "#ffffff",
}

export default colors
