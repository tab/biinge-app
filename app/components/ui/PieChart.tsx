import React from "react"
import { StyleProp, StyleSheet } from "react-native"
import Animated, { FadeIn } from "react-native-reanimated"
import { ProgressCircle, PieChart } from "react-native-svg-charts"

import colors from "styles/colors"

type Props = {
  percent: number
  color?: string
  style?: StyleProp<any>
}

const PieChartComponent = ({ percent, color, style }: Props) => {
  const value = Math.min(100, Math.max(0, percent))

  const data = [
    {
      key: 1,
      value: percent,
      svg: {
        fill: color ? color : colors.white,
      },
      arc: {
        startAngle: 0,
        endAngle: (value * Math.PI) / 50,
      },
    },
    {
      key: 2,
      value: percent,
      svg: {
        fill: "transparent",
      },
    },
  ]

  return (
    <Animated.View entering={FadeIn} style={[style, styles.root]}>
      <ProgressCircle
        style={[styles.chart, styles.circle]}
        progress={1}
        progressColor={colors.white}
        strokeWidth={1}
      />
      <PieChart
        style={[styles.chart, styles.pie]}
        innerRadius={0}
        startAngle={0}
        endAngle={50}
        // @ts-ignore
        data={data}
      />
    </Animated.View>
  )
}

export default PieChartComponent

const styles = StyleSheet.create({
  root: {
    flex: 1,
    height: 20,
    width: 20,
  },
  chart: {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  },
  circle: {
    height: 20,
    width: 20,
  },
  pie: {
    top: 3,
    left: 3,
    height: 14,
    width: 14,
  },
})
