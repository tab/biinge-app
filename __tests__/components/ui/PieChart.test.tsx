import React from "react"
import "react-native"
import { render } from "@testing-library/react-native"
import { it } from "@jest/globals"

import PieChart from "components/ui/PieChart"
import colors from "styles/colors"

describe("PieChart", () => {
  it("renders with default props", () => {
    const tree = render(<PieChart percent={50} />)
    expect(tree).toMatchSnapshot()
  })

  it("renders with custom color", () => {
    const tree = render(<PieChart percent={75} color={colors.orangeYellow} />)
    expect(tree).toMatchSnapshot()
  })

  it("renders with zero percent", () => {
    const tree = render(<PieChart percent={0} />)
    expect(tree).toMatchSnapshot()
  })

  it("renders with 100 percent", () => {
    const tree = render(<PieChart percent={100} />)
    expect(tree).toMatchSnapshot()
  })

  it("renders with over 100 percent (should cap at 100)", () => {
    const tree = render(<PieChart percent={150} />)
    expect(tree).toMatchSnapshot()
  })
})
