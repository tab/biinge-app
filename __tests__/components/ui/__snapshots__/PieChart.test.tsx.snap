// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart renders with 100 percent 1`] = `
<View
  entering={
    {
      "delay": [MockFunction],
    }
  }
  style={
    [
      undefined,
      {
        "flex": 1,
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 14,
          "left": 3,
          "top": 3,
          "width": 14,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`PieChart renders with custom color 1`] = `
<View
  entering={
    {
      "delay": [MockFunction],
    }
  }
  style={
    [
      undefined,
      {
        "flex": 1,
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 14,
          "left": 3,
          "top": 3,
          "width": 14,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`PieChart renders with default props 1`] = `
<View
  entering={
    {
      "delay": [MockFunction],
    }
  }
  style={
    [
      undefined,
      {
        "flex": 1,
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 14,
          "left": 3,
          "top": 3,
          "width": 14,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`PieChart renders with over 100 percent (should cap at 100) 1`] = `
<View
  entering={
    {
      "delay": [MockFunction],
    }
  }
  style={
    [
      undefined,
      {
        "flex": 1,
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 14,
          "left": 3,
          "top": 3,
          "width": 14,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`PieChart renders with zero percent 1`] = `
<View
  entering={
    {
      "delay": [MockFunction],
    }
  }
  style={
    [
      undefined,
      {
        "flex": 1,
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 14,
          "left": 3,
          "top": 3,
          "width": 14,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;
